### PR TITLE
Refactored unit testing

### PR DESCRIPTION
**Pull request by ggsdc**
Adds a new test suite to test the cornflow client api
Adds test to `InstanceSolutionCore` to test the main methods and that they work correctly.
Adds a new workflow on PR to run the integration tests between the client, cornflow and airflow.

The new workflow is pointing to a branch of cornflow server, will open PR there first, then update this PR once that one is closed

### CODE REVIEW

**Comment by ggsdc** [comment]
@pchtsp this PR should as well correct the issue that you created. It adds some tests to the `from_dict`, `to_dict`, `from_json`, `to_json`, `from_excel `and `to_excel` methods.

**Comment by ggsdc** [comment]
This is the current coverage of all the tests:

```
Name                                                             Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------
cornflow_client/__init__.py                                          5      0   100%
cornflow_client/airflow/__init__.py                                  0      0   100%
cornflow_client/airflow/api.py                                     103    103     0%   6-176
cornflow_client/airflow/dag_utilities.py                           101     69    32%   47-51, 61-70, 100-103, 112-131, 135-138, 142, 154-203
cornflow_client/constants.py                                        57     13    77%   95-100, 103-105, 112-115
cornflow_client/core/__init__.py                                     5      0   100%
cornflow_client/core/application.py                                 90     38    58%   43, 52, 62, 72, 82, 96, 106, 131, 138-194, 203, 218
cornflow_client/core/experiment.py                                  26      5    81%   32, 39, 55, 64, 74
cornflow_client/core/instance.py                                    13      7    46%   23, 32, 41-45
cornflow_client/core/instance_solution.py                           72      6    92%   82, 143-145, 158-159
cornflow_client/core/read_tools.py                                  48     24    50%   25-27, 31-33, 62-63, 74-82, 88-92, 94, 97, 99, 110
cornflow_client/core/solution.py                                     4      0   100%
cornflow_client/core/tools.py                                       11      3    73%   16, 26-27
cornflow_client/cornflow_client.py                                 308     28    91%   27, 52, 138, 191, 218, 241-244, 248, 294, 324, 362, 395, 406-413, 429, 450-458, 649-652, 709, 713
cornflow_client/data/__init__.py                                     0      0   100%
cornflow_client/schema/__init__.py                                   0      0   100%
cornflow_client/schema/dictSchema.py                               103      6    94%   87-96, 240, 253
cornflow_client/schema/dict_functions.py                            68      7    90%   30-32, 39, 47, 76, 113
cornflow_client/schema/manager.py                                  116     18    84%   83-84, 106-107, 153, 172, 183-193, 255, 279, 295
cornflow_client/schema/tools.py                                     33     14    58%   46-61, 72, 82
cornflow_client/tests/__init__.py                                    0      0   100%
cornflow_client/tests/const.py                                       9      0   100%
cornflow_client/tests/data/__init__.py                               0      0   100%
cornflow_client/tests/integration/__init__.py                        0      0   100%
cornflow_client/tests/integration/test_cornflow_integration.py     266      0   100%
cornflow_client/tests/unit/__init__.py                               0      0   100%
cornflow_client/tests/unit/test_abc.py                              76      7    91%   69, 74, 80, 85, 88, 91, 96
cornflow_client/tests/unit/test_dag_utilities.py                    15      0   100%
cornflow_client/tests/unit/test_instance_solution_methods.py        81      0   100%
cornflow_client/tests/unit/test_schema_manager.py                  145      0   100%
cornflow_client/tests/unit/test_varnames.py                         81      0   100%
----------------------------------------------------------------------------------------------
TOTAL                                                             1836    348    81%
```

We are still missing specific test in the `airflow api`, although it gets used on internally on cornflow so it is indirectly tested, and some part of the `from_excel` method. Will try to add them later this week. But this should help have a better test suite and bump from the previous 50% coverage.

**Review by marioncottard — APPROVED** [review]
Looks good to me!

**Comment by ggsdc** [comment]
> Looks good to me!

Ok I'm going to make some more tests for the airflow api and then you can review the last changes

File: cornflow_client/tests/const.py [hunk 1]
```diff
@@ -79,3 +86,20 @@
...
+    "bar_cutting",
+    "facility_location",
+    "graph_coloring",
+    "hk_2020_dag",
+    "knapsack",
+    "roadef",
+    "rostering",
+    "tsp",
+    "vrp",
+]
```
Comment (pchtsp) [inline cornflow_client/tests/const.py:105]: are you going to update this list as the public repo is updated? Sounds ambitious.

File: cornflow_client/core/instance_solution.py [hunk 2]
```diff
@@ -146,9 +146,11 @@ def to_excel(self, path: str):
 
         is_xl_type(path)
```
Comment (pchtsp) [inline cornflow_client/core/instance_solution.py:147]: I'm curious about this function. Does it raise an error if the path is not excel? Because from the name it looks as if it just returns True or False.
Also, above I see a warning *and* an exception being razed? Does it make sense to do both?

**Review by pchtsp — COMMENTED** [review]
I'm not 100% convinced on executing cornflow server and airflow to test this library... Couldn't the testing be done by mocking the REST service? That way we can test cornflow-client without actually having to deploy the server in github actions. With the added benefit that if someone wants to run the tests of this library, they will able to do so without having to deploy the other two (and have linux). Also, most of the integration tests are already being done in the cornflow-server repo since it uses the client to test.

**Comment by ggsdc** [comment]
> I'm not 100% convinced on executing cornflow server and airflow to test this library... Couldn't the testing be done by mocking the REST service? That way we can test cornflow-client without actually having to deploy the server in github actions. With the added benefit that if someone wants to run the tests of this library, they will able to do so without having to deploy the other two (and have linux). Also, most of the integration tests are already being done in the cornflow-server repo since it uses the client to test.

I've been having problems trying to mock the behaviour of cornflow-server and airflow, that is why I created this tests. As it is a public repository there is no restrictions on the actions so there is no problem about limits.

If someone where to fork this library, they could run the unit test on their local machine and leave the integration test to do it on the github actions of their fork.

I know that the library is getting indirectly tested on the cornflow-server side, but that would not allow to detect a change that would broke something when just working on this directory, unless you went to the cornflow-server repo, changed the branch of the cornflow-client on the requirements.txt file to point to their development branch of cornflow-client and check that everything works as intended, which I think is a worst workflow that to have the test directly here.

Having the test here allows us to check if a change in this repository breaks anything without having to run the test on cornflow-server.

Ideally I would like to have images of cornflow-server and airflow somewhere so the workflow just imports the images and is faster and easier to run, as those images could be run on systems that are not linux with the help of docker.

So in summary, I agree this is not the best solution, but I do not think is a bad one either, it lets us check any changes without having to run the test on cornflow-server, it gets run automatically on PR and using the real server and not a mock one, makes the test have less maintenance, as with a mock testing any change in cornflow-server would need to review that the mock behaviour here is the same.

(sorry for the long reply, I'm more that happy to talk about this more in detail in a call if you want to).

File: cornflow_client/tests/const.py (hunk 1 above)
Comment (ggsdc) [reply cornflow_client/tests/const.py:105]: This list is the list of the current dags on the public repo. The test checks that at least these dags are deployed, so if add dags to the public repo the test here would still pass.

The only problem would be if we change the name of a dag in the public repo or if we delete one of them, cases that I think are going to be rare, but anyway, I will try to think of any other solution.

File: cornflow_client/core/instance_solution.py (hunk 2 above)
Comment (ggsdc) [reply cornflow_client/core/instance_solution.py:147]: It does raise an exception: 

```
if not any(ext in path for ext in [".xlsx", ".xls", ".xlsm"]):
        raise ConnectionError(
            "Excel_file argument should be a string with an excel extension"
        )
```

And yes you are right we do not need both a warning and an exception on module import. Will take one of them out.

**Comment by ggsdc** [comment]
I'm going to merge this for the time being until we get the monorepo up and running on the main cornflow repository. Until then I think is a good idea to perform integration tests here to ensure that any change here does not break the current use of the cornflow client library on cornflow-server
